spec/lib/scheduler/thread_pool_spec.rb

```
@@ -0,0 +1,164 @@
+      current_db = "default"
+      completion_queue = Queue.new
+
+      allow(RailsMultisite::ConnectionManagement).to receive(:current_db).and_return(current_db)
+      allow(RailsMultisite::ConnectionManagement).to receive(:with_connection).with(
+        current_db,
+      ) do |&block|
+        completion_queue << current_db
+        block.call
+      end
+
+      pool.post { true }
+      expect(completion_queue.pop).to eq(current_db)
+    end
+
+    it "scales up threads when work increases" do
+      completion_queue = Queue.new
+      blocker_queue = Queue.new
+
+      # Create enough blocking tasks to force thread creation
+      (max_threads + 1).times do |i|
+        pool.post do
+          completion_queue << i
+          blocker_queue.pop
+        end
+      end
+
+      expect(pool.stats[:active_threads]).to eq(max_threads)
+      (max_threads + 1).times { blocker_queue << :continue }
+
+      results = Array.new(max_threads + 1) { completion_queue.pop }
+
+      expect(results.sort).to eq((0..max_threads).to_a)
+    end
+  end
+
+  describe "#shutdown" do
+    it "prevents new tasks from being posted" do
+      completion_queue = Queue.new
+      pool.post { completion_queue << 1 }
+      completion_queue.pop # ensure first task completes
+
+      pool.shutdown
+      expect(pool.shutdown?).to be true
+      expect { pool.post { true } }.to raise_error(Scheduler::ThreadPool::ShutdownError)
+    end
+
+    it "completes pending tasks before shutting down" do
+      completion_queue = Queue.new
+      3.times { |i| pool.post { completion_queue << i } }
+
+      results = Array.new(3) { completion_queue.pop }
+      pool.shutdown
+
+      expect(results.size).to eq(3)
+      expect(results.sort).to eq([0, 1, 2])
+    end
+  end
+
+  describe "error handling" do
+    it "captures and logs exceptions without crashing the thread" do
+      completion_queue = Queue.new
+      error_msg = "Test error"
+
+      pool.post { raise StandardError, error_msg }
```

I think it can be useful to test that the `error_msg` was logged. If we don't have this assertion, we might end up having work silently failing in the thread pool

---

that means we need a test logger hmm thinking about it.